pkg/security/securitycontextconstraints/provider.go

```diff
@@ -292,19 +292,38 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
 
 	allErrs = append(allErrs, s.capabilitiesStrategy.Validate(pod, container)...)
 
-	if len(pod.Spec.Volumes) > 0 && !sccutil.SCCAllowsAllVolumes(s.scc) {
-		allowedVolumes := sccutil.FSTypeToStringSet(s.scc.Volumes)
-		for i, v := range pod.Spec.Volumes {
-			fsType, err := sccutil.GetVolumeFSType(v)
-			if err != nil {
-				allErrs = append(allErrs, field.Invalid(fldPath.Child("volumes").Index(i), string(fsType), err.Error()))
-				continue
+	if len(pod.Spec.Volumes) > 0 {
+		if !sccutil.SCCAllowsAllVolumes(s.scc) {
+			allowedVolumes := sccutil.FSTypeToStringSet(s.scc.Volumes)
+			for i, v := range pod.Spec.Volumes {
+				fsType, err := sccutil.GetVolumeFSType(v)
+				if err != nil {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("volumes").Index(i), string(fsType), err.Error()))
+					continue
+				}
+
+				if !allowedVolumes.Has(string(fsType)) {
+					allErrs = append(allErrs, field.Invalid(
+						fldPath.Child("volumes").Index(i), string(fsType),
+						fmt.Sprintf("%s volumes are not allowed to be used", string(fsType))))
+				}
 			}
+		}
 
-			if !allowedVolumes.Has(string(fsType)) {
-				allErrs = append(allErrs, field.Invalid(
-					fldPath.Child("volumes").Index(i), string(fsType),
-					fmt.Sprintf("%s volumes are not allowed to be used", string(fsType))))
+		if len(s.scc.AllowedFlexDrivers) > 0 && sccutil.SCCAllowsFSType(s.scc, securityapi.FSTypeFlexVolume) {
```

Comment:
if the volume type is completely disallowed, line 306 would already have added an error. there's no need to add another error saying the specific flex driver is not allowed. making a single loop over all the volumes would let you short-circuit adding specific errors if there are general ones.

```
allowAllVolumeTypes := sccutil.SCCAllowsAllVolumes(s.scc)
allowedVolumeTypes := sccutil.FSTypeToStringSet(s.scc.Volumes)
for i, v := range pod.Spec.Volumes {
  fsType, err := sccutil.GetVolumeFSType(v)
  if err != nil {
    ... add the error...
    continue
  }

  if !allowAllVolumeTypes && !allowedVolumes.Has(string(fsType)) {
    ... add an error about this volume type not being allowed...
    continue
  }

  if v.FlexVolume != nil && len(s.scc.AllowedFlexDrivers) > 0 {
    ... ensure the particular flex volume is allowed...
  }
}
```